arangod/RestServer/QueryRegistryFeature.cpp

```diff
@@ -794,6 +832,42 @@ void QueryRegistryFeature::unprepare() {
   QUERY_REGISTRY.store(nullptr, std::memory_order_release);
 }
 
+size_t QueryRegistryFeature::leaseAsyncPrefetchSlots(size_t value) noexcept {
+  // reduce value to the allowed per-query maximum
+  value = std::min(_maxAsyncPrefetchSlotsPerQuery, value);
+
+  if (value == 0) {
+    return 0;
+  }
+
+  // check if global (per-server) total is below the configured total maximum
+  size_t expected = _asyncPrefetchSlotsUsed.load(std::memory_order_relaxed);
+  TRI_ASSERT(expected <= _maxAsyncPrefetchSlotsTotal);
+  do {
+    size_t target = std::min(_maxAsyncPrefetchSlotsTotal, expected + value);
+    if (target == expected) {
+      // no slots left
+      return 0;
+    }
+    TRI_ASSERT(target > expected);
+    if (_asyncPrefetchSlotsUsed.compare_exchange_weak(
+            expected, target, std::memory_order_relaxed,
+            std::memory_order_relaxed)) {
+      return target - expected;
+    }
+    TRI_ASSERT(expected <= _maxAsyncPrefetchSlotsTotal);
+  } while (true);
```

Comment:
I would like to reformulate/simplify this a bit:
```suggestion
  size_t expected = _asyncPrefetchSlotsUsed.load(std::memory_order_relaxed);
  TRI_ASSERT(expected <= _maxAsyncPrefetchSlotsTotal);
  // as long as we have slots left, try to allocate some
  while (expected < _maxAsyncPrefetchSlotsTotal) {
    size_t target = std::min(_maxAsyncPrefetchSlotsTotal, expected + value);
    TRI_ASSERT(target > expected);
    if (_asyncPrefetchSlotsUsed.compare_exchange_weak(
            expected, target, std::memory_order_relaxed)) {
      return target - expected;
    }
    TRI_ASSERT(expected <= _maxAsyncPrefetchSlotsTotal);
  };
```